File: 2023/04.24/서지훈_백준_2812_크게_만들기.java
import java.io.*;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        int k = K;

        String num = br.readLine();

        Stack<Integer> stack = new Stack<>();

        for(int i = 0; i < num.length(); ++i){
            if(stack.isEmpty()){
                stack.push(num.charAt(i) - '0');
            }
            else if(!stack.isEmpty() && stack.peek() >= num.charAt(i) - '0'){
                stack.push(num.charAt(i) - '0');
            }
            else{
                while(!stack.isEmpty() && K > 0 && stack.peek() < num.charAt(i) - '0'){
                    stack.pop();
                    K -= 1;
                }
                stack.push(num.charAt(i) - '0');
            }
        }

        int cnt = 0;
        for(Integer ele : stack){
            if(cnt++ < N - k){
                bw.write(Integer.toString(ele));
            }
        }

        bw.flush();
        bw.close();
        br.close();
    }
}

File: 2023/04.24/서지훈_백준_2812_크게_만들기.cpp
#include <stack>
#include <iostream>

int main(){
	int N, K;
	std::string str;
	
	std::cin >> N >> K;
	std::cin >> str;
	
	int k = K;
	
	std::stack<char> st;
	
	for(int i = 0; i < str.length(); ++i){
		if(st.empty()){
			st.push(str.at(i));
		}
		else if(K > 0 && st.top() >= str.at(i)){
			st.push(str.at(i));
		}
		else{
			while(!st.empty() && K > 0 && st.top() < str.at(i)){
				st.pop();
				K -= 1;
			}
			st.push(str.at(i));
		}
	}
	
	std::stack<char> st2;
	while(!st.empty()){
		st2.push(st.top());
		st.pop();
	}
	int cnt = 0;
	while(cnt++ < N - k){
		std::cout << st2.top();
		st2.pop();
	}
	
}